runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1862,7 +1861,15 @@ J9::Z::PrivateLinkage::buildVirtualDispatch(TR::Node * callNode, TR::RegisterDep
 
             if (useProfiledValues)
                {
-               TR::Instruction * unloadableConstInstr = generateRILInstruction(cg(), TR::InstOpCode::LARL, callNode, RegZero, reinterpret_cast<uintptr_t*>(profiledClass));
+               TR::Instruction * unloadableConstInstr;
+               if (cg()->canUseRelativeLongInstructions(reinterpret_cast<uintptr_t>(profiledClass)))
+                  {
+                  unloadableConstInstr = generateRILInstruction(cg(), TR::InstOpCode::LARL, callNode, RegZero, reinterpret_cast<uintptr_t*>(profiledClass));
+                  }
+               else
+                  {
+                  unloadableConstInstr = genLoadAddressConstantInSnippet(cg(), callNode, reinterpret_cast<uintptr_t>(profiledClass), RegZero, NULL, dependencies, NULL, true);
```

Comment:
Using `genLoadAddressConstantInSnippet` here is causing problems with class unloading. When the profiled class is unloaded, the constant data snippet holding the class address is not patched.

@r30shah and I did some investigation into why this is happening.
[`genLoadAddressConstantInSnippet`][1] calls [`generateRegLitRefInstruction`][2] to create the constant data snippet to hold the address. `generateRegLitRefInstruction` calls [`rememberSnippetsToBePatchedOnClassUnloadOrRedefinition`][3] which should mark the snippet to be patched on class unload. However, before adding the snippet to the list of snippets that need to be patched on class unload, `rememberSnippetsToBePatchedOnClassUnloadOrRedefinition` checks [`isClassUnloadingConst`][4] on the node, which can only return true for `aloadi` and `aconst` nodes, not for a virtual call node like we would have here in devirtualization.

One possible workaround is to directly add the snippet to list of snippets to be patched here:
```diff
@@ -1869,6 +1869,9 @@ J9::Z::PrivateLinkage::buildVirtualDispatch(TR::Node * callNode, TR::RegisterDep
                else
                   {
                   unloadableConstInstr = genLoadAddressConstantInSnippet(cg(), callNode, reinterpret_cast<uintptr_t>(profiledClass), RegZero, NULL, dependencies, NULL, true);
+                  cg()->getSnippetsToBePatchedOnClassUnload()->push_front(static_cast<TR::S390RILInstruction *>(unloadableConstInstr)->getTargetSnippet());
                   }
                if (fej9->isUnloadAssumptionRequired(profiledClass, comp()->getCurrentMethod()))
                   {
```

[1]: https://github.com/eclipse/omr/blob/cd8a2e017030701fc5b39ea162acdec128c6aecb/compiler/z/codegen/OMRTreeEvaluator.cpp#L2220-L2225
[2]: https://github.com/eclipse/omr/blob/cd8a2e017030701fc5b39ea162acdec128c6aecb/compiler/z/codegen/S390GenerateInstructions.cpp#L2537
[3]: https://github.com/eclipse/omr/blob/cd8a2e017030701fc5b39ea162acdec128c6aecb/compiler/z/codegen/S390GenerateInstructions.cpp#L2337
[4]: https://github.com/eclipse/omr/blob/cd8a2e017030701fc5b39ea162acdec128c6aecb/compiler/il/OMRNode.cpp#L3021

Reply:
A bit of info on how we confirmed that the snippet was not patched (also how we confirmed that the snippet *is* patched when directly added to `SnippetsToBePatchedOnClassUnload`). We used this test:

```java
mport java.io.File;
import java.net.URL;
import java.net.URLClassLoader;
import java.net.MalformedURLException;
import java.lang.reflect.Field;

public class UnitTest {
        public static int unitTest(Parent obj, int par) {
                return obj.method(par);
        }   

        public static void main(String[] args) throws MalformedURLException, ClassNotFoundException, InstantiationException, IllegalAccessException, I
                File file = new java.io.File("classloader1");
                ClassLoader loader = new URLClassLoader(new URL[] {file.toURL()});
                Class<?> child1Clz = loader.loadClass("Child1");
                Class<?> child2Clz = loader.loadClass("Child2");
                Class<?> grandChild3Clz = loader.loadClass("GrandChild3");
                Class<?> grandChild2Clz = loader.loadClass("GrandChild2");

                for (int i=0; i < 1000; i++) {
                        Object obj = null;
                        if (i%2 == 0) {
                                obj = child1Clz.newInstance();
                        } else if (i%3 == 0) {
                                obj = child2Clz.newInstance();
                        } else if (i%5 == 0) {
                                obj = grandChild3Clz.newInstance();
                        } else {
                                obj = grandChild2Clz.newInstance();
                        }   
                        ((Parent)obj).method(i);
                }   
                for (int i=0; i < 100000; i++) {
                        Object obj = null;
                        if (i%2 == 0) {
                                obj = child1Clz.newInstance();
                        } else if (i%3 == 0) {
                                obj = child2Clz.newInstance();
                        } else if (i%5 == 0) {
                                obj = grandChild3Clz.newInstance();
                        } else {
                                obj = grandChild2Clz.newInstance();
                        }   
                        unitTest((Parent)obj, i); 
                }   

                System.out.println("Sleeping for 30s");
                Thread.sleep(30000);

                System.out.println("Before unloading loading class from ClassLoader1, printing static invocation count from Child1");
                Field f = child1Clz.getDeclaredField("invocationCount");
                f.setAccessible(true);
                System.out.println((int)f.get(null));
                child1Clz = null;
                child2Clz = null;
                grandChild3Clz = null;
                grandChild2Clz = null;
                loader = null;

                System.out.println("Forcing GC");
                for (int i=0; i < 3; i++) {
                        System.gc();
                        System.gc();
                        Thread.sleep(1000);
                }   
                System.out.println("Sleeping for 30s");
                Thread.sleep(30000);

                file = new java.io.File("classloader2");
                loader = new URLClassLoader(new URL[] {file.toURL()});
                child1Clz = loader.loadClass("Child1");
                child2Clz = loader.loadClass("Child2");
                grandChild3Clz = loader.loadClass("GrandChild3");
                grandChild2Clz = loader.loadClass("GrandChild2");
                for (int i=0; i < 100000; i++) {
                        Object obj = null;
                        if (i%2 == 0) {
                                obj = child1Clz.newInstance();
                        } else if (i%3 == 0) {
                                obj = child2Clz.newInstance();
                        } else if (i%5 == 0) {
                                obj = grandChild3Clz.newInstance();
                        } else {
                                obj = grandChild2Clz.newInstance();
                        }   
                        unitTest((Parent)obj, i); 
                }   
                System.out.println("Before unloading loading class from ClassLoader2, printing static invocation count from Child1");
                Field f1 = child1Clz.getDeclaredField("invocationCount");
                f1.setAccessible(true);
                System.out.println((int)f1.get(null));
        }   
}
```

```bash
java -Xjit:"count=100,disableAsyncCompilation,dontInline={*unitTest*},{*unitTest*}(count=200,disableInlining,traceFull,log=methodLog),verbose,vlog=jit.verbose" "-Xdump:system+java:events=user,request=exclusive+prepwalk+preempt" UnitTest
```

During each of the two 30 second sleeps, I ran `pkill -3 java` to send `SIGQUIT` to the program, triggering a core dump. This way we could compare the core dumps from before and after the class unloading. With the compiled method address from the compilation log, we found the generated snippet in each core dump. When I did this without directly adding the snippet to be patched, the generated snippets were identical. When did this with the snippet directly added, the generated snippet was patched with `ffff ffff ffff ffff` in the core dump taken after the class unload.